### ISSUE TITLE
Bad particle output in parallel runs

### ISSUE THREAD

**Issue by bartniday** [geodynamics/aspect#2572]
When running parallel Aspect models that use particles, I've been getting vtk output files that immediately crash Paraview when loaded (trying hdf5 output instead, Aspect has a segmentation fault during postprocessing).

It looks like the problem is something to do with the particle distribution. I'm running in chunk geometry, and the models work fine when the particle generator is set to 'random uniform' or to 'uniform radial' with the limits set to the whole model extent. But if I try to use 'uniform radial' with limits set to cover a (small) subregion of the model, the output files do not work - Aspect runs normally and produces vtk files, but paraview crashes on loading them. Changing for example the number of particles doesn't seem to affect this.

**Comment by gassmoeller** [comment]
Hi Bart, that sounds like it might be a bug indeed. Can you attach a parameter file that reproduces the problem (ideally small enough to run it on a laptop)? My first guess would be that the files on processes without particles might be corrupted. You can trace this down by trying to open individual solution-xxxx.xxxx.vtu files, and see which one of those does not open. You can then open the corrupted files in a text editor and see if the structure looks similar compared to another file that opens correctly in Paraview. Let me know what you find (or attach a .prm file if you want me to take a look).

**Comment by bartniday** [comment]
I haven't yet been able to reproduce this problem on my own computer - running the same parameter file with fewer refinements

**Comment by bartniday** [comment]
Reproducing the issue on my desktop computer (although running a different version of Aspect) hasn't been consistent. Sometimes the same parameter file will run on my own computer but not on the cluster with the same number of processes.

The attached .prm file is the best I've done so far. It crashes with a segmentation fault when I run it on the cluster with two processes. If I change the "radial layers" parameter in `Postprocess/Particles/Generator/Uniform radial` from 10 to 1, it works normally. If I change particle output from hdf5 to vtu, it also works normally.
The parameter file also runs normally on my office computer, but my office computer is using older versions of Aspect and the associated libraries.


[particle_segmentation_fault.prm.txt](https://github.com/geodynamics/aspect/files/2203059/particle_segmentation_fault.prm.txt)

**Comment by bartniday** [comment]
Hi Rene,

After getting back to this issue, I tried those changes and they don't seem to fix the problem with .vtu files.

I attached a .prm file that makes broken .vtu files when run on my cluster with 28 processors. The file works normally on my laptop, or if I run it on the cluster with 4 processors. I also attached one of the offending .vtu files.

[bad-vtk-output.prm.txt](https://github.com/geodynamics/aspect/files/2250952/bad-vtk-output.prm.txt)
[particles-00000.0000.vtu.txt](https://github.com/geodynamics/aspect/files/2250948/particles-00000.0000.vtu.txt)

**Comment by gassmoeller** [comment]
Hi Bart,
can you confirm for me that everything works, if you add the following into the `Particles` subsection?
```
set Number of grouped files = 0
```

It seems Paraview can not handle the case when there are several 'pieces' (from different MPI processes) in the same file, and one of those pieces has 0 points (as can happen if a single process has no particles). I am not sure how to fix this best, in the worst case we need to disable grouping of files for particle output. Just not writing pieces with 0 points does not work, because each file needs to contain a valid header and footer, and at least one piece (which may contain 0 points, but not if there is another piece with more than 0 points). This is not only a paraview issue, as visit does not crash when reading these file, but instead just not loads any data (and if I open the .pvtu it shows only data from files that do not contain pieces with 0 points). Not sure what to do yet, but the suggestion above should at least provide a workaround. Let me know if it works.

**Comment by bartniday** [comment]
Adding that line fixes the vtk output for me (in my models as well as in the example .prm file).

**Comment by bangerth** [comment]
@gassmoeller -- do I understand you right that the presence of this block here is what causes the trouble:
```
<Piece NumberOfPoints="0" NumberOfCells="0" >
<Cells>
<DataArray type="UInt8" Name="types"></DataArray>
</Cells>
  <PointData Scalars="scalars">
    <DataArray type="Float32" Name="initial position" NumberOfComponents="3"></DataArray>
    <DataArray type="Float32" Name="id"></DataArray>
  </PointData>
</Piece>
```
If you remove this block from the VTU file, does it load correctly? If we assume that all simulations have at least one non-empty block, could we not just omit the empty blocks?

Looking at the input file, I also notice that the non-empty "piece" has more `<DataArray>` entries. Similarly, it has a `<Points>` block, whereas the empty one does have that. Should they be there, but empty?

**Comment by gassmoeller** [comment]
Yes this block is the problem. And no, we can not simply omit writing every block with 0 points, because each file needs to contain at least one block, and that block can be empty (think of 4 processes, only one of the processes has particles, and we write into 2 files, one file would need to contain one empty block, the other one would need to contain one block with the actual output).

The observation about the different `<DataArray>` entries is very promising. I copied the missing entries from the piece with points into the piece without points (I had to remove the data, and also the `format="binary"` part in the header), and then Paraview loads the file without problems. So it looks like we need to write empty blocks with the same data arrays as non-empty blocks to make this work.
